File: src/main/java/nl/openweb/quarkus/samples/petclinic/Resource.java
package nl.openweb.quarkus.samples.petclinic;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;

import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import java.util.List;

@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public interface Resource<T extends PanacheEntityBase> {

    @GET
    List<T> getAll();

    @GET
    @Path("{id}")
    T getById(@PathParam("id") long id);

    @POST
    T create(T restEntity);

    @PUT
    @Path("{id}")
    T update(@PathParam("id") long id, T restEntity);

    @PUT
    @Path("{id}")
    void delete(@PathParam("id") long id);
}

File: src/main/kotlin/nl/openweb/quarkus/Resource.kt
package nl.openweb.quarkus

import io.quarkus.hibernate.orm.panache.PanacheRepository
import javax.ws.rs.Consumes
import javax.ws.rs.DELETE
import javax.ws.rs.GET
import javax.ws.rs.POST
import javax.ws.rs.PUT
import javax.ws.rs.Path
import javax.ws.rs.PathParam
import javax.ws.rs.Produces
import javax.ws.rs.core.MediaType

@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
abstract class Resource<T>(
   private val repository: PanacheRepository<T>
) {
    @GET
    @Path("")
    fun getAll(): List<T> = repository.listAll()

    @GET
    @Path("{id}")
    fun getById(@PathParam("id") id: Long): T = repository.findById(id)

    @POST
    @Path("")
    open fun create(restEntity: T) = restEntity.apply { repository.persist(this) }

    @PUT
    @Path("{id}")
    abstract fun update(@PathParam("id") id: Long, restEntity: T): T

    @DELETE
    @Path("{id}")
    open fun delete(@PathParam("id") id: Long) = repository.delete(getById(id))
}
